File: ios/ZenCompanionTests/App/DemoCatalogTests.swift
import XCTest
@testable import ZenCompanion

final class DemoCatalogTests: XCTestCase {
    func testSnapshotCoversCoreSurfaces() {
        let snap = DemoCatalog.snapshot
        XCTAssertGreaterThanOrEqual(snap.spaces.count, 3)
        XCTAssertTrue(snap.spaces.contains { !$0.pinned.isEmpty })
        XCTAssertTrue(snap.spaces.contains { !$0.tabs.isEmpty })
        XCTAssertTrue(snap.spaces.contains { space in
            space.pinned.contains { item in
                if case .folder = item { return true }
                return false
            }
        })
        XCTAssertTrue(snap.spaces.contains { space in
            space.pinned.contains { item in
                if case .split = item { return true }
                return false
            }
        })
        XCTAssertEqual(snap.essentials.count, snap.spaces.count)
        XCTAssertTrue(snap.spaces.contains { space in
            space.pinned.contains { item in
                if case .folder(let folder) = item {
                    return !(folder.subfolders ?? []).isEmpty
                }
                return false
            }
        })
    }

    func testActivityHasHistory() {
        XCTAssertFalse(DemoCatalog.activity.history.isEmpty)
    }

    func testLegacyAccountDecodesWithoutDemoFlag() throws {
        let json = """
        {"email":"a@b.c","uid":"u","sessionTokenHex":"aa","kBHex":"bb"}
        """.data(using: .utf8)!
        let snap = try JSONDecoder().decode(AccountSnapshot.self, from: json)
        XCTAssertFalse(snap.isDemo)
        XCTAssertEqual(snap.email, "a@b.c")
    }

    func testDemoAccountRoundTrips() throws {
        let data = try JSONEncoder().encode(DemoCatalog.account)
        let snap = try JSONDecoder().decode(AccountSnapshot.self, from: data)
        XCTAssertTrue(snap.isDemo)
        XCTAssertEqual(snap.uid, "demo")
    }
}

final class FxASocialAuthTests: XCTestCase {
    func testBlocksAppleAndGoogleAuthHosts() {
        XCTAssertTrue(FxASocialAuth.shouldBlock(URL(string: "https://appleid.apple.com/auth/authorize")!))
        XCTAssertTrue(FxASocialAuth.shouldBlock(URL(string: "https://accounts.google.com/o/oauth2/v2/auth")!))
    }

    func testAllowsMozillaAccounts() {
        XCTAssertFalse(FxASocialAuth.shouldBlock(URL(string: "https://accounts.firefox.com/?service=sync")!))
        XCTAssertFalse(FxASocialAuth.shouldBlock(URL(string: "https://api.accounts.firefox.com/v1/account/status")!))
    }
}

File: android/app/src/test/java/de/kjell/zencompanion/DemoCatalogTests.kt
package de.kjell.zencompanion

import de.kjell.zencompanion.data.AccountStore
import de.kjell.zencompanion.data.DemoCatalog
import de.kjell.zencompanion.sync.ZenSpaces
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/** Ports of `DemoCatalogTests.swift`. */
class DemoCatalogTests {
    @Test
    fun snapshotCoversCoreSurfaces() {
        val snap = DemoCatalog.snapshot
        assertTrue(snap.spaces.size >= 3)
        assertTrue(snap.spaces.any { it.pinned.isNotEmpty() })
        assertTrue(snap.spaces.any { it.tabs.isNotEmpty() })
        assertTrue(
            snap.spaces.any { space ->
                space.pinned.any { it is ZenSpaces.ZenItem.Folder }
            },
        )
        assertTrue(
            snap.spaces.any { space ->
                space.pinned.any { it is ZenSpaces.ZenItem.Split }
            },
        )
        assertEquals(snap.spaces.size, snap.essentials.size)
        assertTrue(
            snap.spaces.any { space ->
                space.pinned.any { item ->
                    item is ZenSpaces.ZenItem.Folder &&
                        !(item.folder.subfolders ?: emptyList()).isEmpty()
                }
            },
        )
    }

    @Test
    fun activityHasHistory() {
        assertFalse(DemoCatalog.activity.history.isEmpty())
    }

    @Test
    fun legacyAccountDecodesWithoutDemoFlag() {
        val json = JSONObject(
            """{"email":"a@b.c","uid":"u","sessionTokenHex":"aa","kBHex":"bb"}""",
        )
        val snap = AccountStore.AccountSnapshot.fromJSON(json)!!
        assertFalse(snap.isDemo)
        assertEquals("a@b.c", snap.email)
    }

    @Test
    fun demoAccountRoundTrips() {
        val snap = AccountStore.AccountSnapshot.fromJSON(DemoCatalog.account.toJSON())!!
        assertTrue(snap.isDemo)
        assertEquals("demo", snap.uid)
    }

    @Test
    fun snapshotCacheKeepsNestedFolders() {
        val encoded = ZenSpaces.encodeSnapshot(DemoCatalog.snapshot)
        val decoded = ZenSpaces.decodeSnapshot(encoded)!!
        val nested = decoded.spaces
            .flatMap { it.pinned }
            .filterIsInstance<ZenSpaces.ZenItem.Folder>()
            .flatMap { it.folder.subfolders ?: emptyList() }
        assertTrue(nested.any { it.id == "demo-folder-specs" })
        assertEquals(3, decoded.essentials.size)
    }
}
